File: standardizer.py
import openpyxl
from openpyxl.utils import get_column_letter
from difflib import SequenceMatcher
import re

class ReportStandardizer:
    def __init__(self, target_template_path="Full_Financial_Report_Template.xlsx"):
        self.template_path = target_template_path
        
    def normalize_string(self, s):
        if not s: return ""
        return str(s).strip().replace(" ", "").replace("\u3000", "")

    def fuzzy_match(self, source, target, threshold=0.6):
        source = self.normalize_string(source)
        target = self.normalize_string(target)
        if not source or not target: return False
        return SequenceMatcher(None, source, target).ratio() > threshold

    def standardize(self, input_wb):
        # Create a new workbook based on the template structure (conceptually)
        # For simplicity, we create a new WB and manually populate the known structure
        output_wb = openpyxl.Workbook()
        
        # We need to find data for:
        # 1. Income Statement: [Item, LastYear, Budget, ThisYear, Note]
        # 2. Balance Sheet: [Item, LastYear, ThisYear, Note]
        
        self._process_income_statement(input_wb, output_wb)
        self._process_balance_sheet(input_wb, output_wb)
        self._process_property_catalog(input_wb, output_wb)
        
        return output_wb

    def _find_sheet_by_keyword(self, wb, keywords):
        for sheet_name in wb.sheetnames:
            norm_name = self.normalize_string(sheet_name)
            if any(k in norm_name for k in keywords):
                return wb[sheet_name]
        return None

    def _extract_data(self, ws, required_cols):
        # Intelligent column mapping
        # required_cols = {"this_year": ["本年", "決算"], "last_year": ["上年", "去年"], "budget": ["預算"], "item": ["科目", "摘要"]}
        
        header_row_idx = -1
        col_mapping = {} # key (e.g. 'this_year') -> col_index (1-based)

        # 1. Find Header Row (scan first 10 rows)
        for r in range(1, min(11, ws.max_row + 1)):
            row_values = [str(cell.value or "") for cell in ws[r]]
            matches = 0
            temp_mapping = {}
            
            for key, keywords in required_cols.items():
                for col_idx, cell_val in enumerate(row_values, 1):
                    norm_val = self.normalize_string(str(cell_val))
                    if any(kw in norm_val for kw in keywords):
                        temp_mapping[key] = col_idx
                        break
            
            if len(temp_mapping) >= 2: # At least find Item and one amount
                header_row_idx = r
                col_mapping = temp_mapping
                break
        
        if header_row_idx == -1:
            print(f"Warning: Could not identify headers in sheet {ws.title}")
            return []

        # 2. Extract Data
        extracted_data = [] # List of dicts {key: value}
        for r in range(header_row_idx + 1, ws.max_row + 1):
            row_data = {}
            has_data = False
            
            # Get Item Name
            if "item" in col_mapping:
                item_name = ws.cell(row=r, column=col_mapping["item"]).value
                if item_name:
                    row_data["item"] = item_name
                    has_data = True
            
            if not has_data: continue

            # Get Amounts
            for key, col_idx in col_mapping.items():
                if key == "item": continue
                val = ws.cell(row=r, column=col_idx).value
                # Clean number
                try:
                    if isinstance(val, (int, float)):
                        row_data[key] = val
                    elif val and isinstance(val, str):
                        # Remove commas, handle parenthesis as negative
                        clean_val = val.replace(",", "").strip()
                        if clean_val.startswith("(") and clean_val.endswith(")"):
                            clean_val = "-" + clean_val[1:-1]
                        row_data[key] = float(clean_val)
                    else:
                        row_data[key] = 0
                except:
                    row_data[key] = 0
            
            extracted_data.append(row_data)
        
        return extracted_data

    def _process_income_statement(self, input_wb, output_wb):
        ws_in = self._find_sheet_by_keyword(input_wb, ["收支", "損益", "決算"])
        ws_out = output_wb.create_sheet("收支決算表")
        ws_out.append(["科目名稱", "上年度決算數", "本年度預算數", "本年度決算數", "說明"]) # Standard Header
        
        if not ws_in:
            return

        data = self._extract_data(ws_in, {
            "item": ["科目", "項目", "摘要", "名稱"],
            "this_year": ["本年", "本期", "決算"],
            "last_year": ["上年", "上期", "去年", "比較"],
            "budget": ["預算"]
        })

        for row in data:
            ws_out.append([
                row.get("item", ""),
                row.get("last_year", 0),
                row.get("budget", 0),
                row.get("this_year", 0),
                "AI轉檔"
            ])

    def _process_balance_sheet(self, input_wb, output_wb):
        ws_in = self._find_sheet_by_keyword(input_wb, ["資產", "負債", "平衡"])
        ws_out = output_wb.create_sheet("資產負債表")
        ws_out.append(["科目名稱", "上年度金額", "本年度金額", "說明"])
        
        if not ws_in:
            return

        data = self._extract_data(ws_in, {
            "item": ["科目", "項目", "摘要", "名稱"],
            "this_year": ["本年", "本期", "期末", "決算"],
            "last_year": ["上年", "上期", "期初", "去年"]
        })

        for row in data:
            ws_out.append([
                row.get("item", ""),
                row.get("last_year", 0),
                row.get("this_year", 0),
                "AI轉檔"
            ])
            
    def _process_property_catalog(self, input_wb, output_wb):
         # Create empty or try to find
         ws_out = output_wb.create_sheet("財產目錄")
         ws_out.append(["資產名稱", "取得日期", "取得成本", "本期折舊", "帳面價值"])
         # Simple copy logic omitted for brevity, just ensuring sheet exists

File: js/standardizer.js
export class ReportStandardizer {
    constructor() {
        // Keywords for identifying sheets and columns
    }

    normalizeString(s) {
        if (!s) return "";
        return String(s).trim().replace(" ", "").replace("\u3000", "");
    }

    standardize(inputWb) {
        // Create a new workbook
        const outputWb = XLSX.utils.book_new();

        this._processIncomeStatement(inputWb, outputWb);
        this._processBalanceSheet(inputWb, outputWb);
        this._processFundStatement(inputWb, outputWb); // New
        this._processPropertyCatalog(inputWb, outputWb);

        return outputWb;
    }

    _findSheetByKeyword(wb, keywords) {
        for (const sheetName of wb.SheetNames) {
            const normName = this.normalizeString(sheetName);
            if (keywords.some(k => normName.includes(k))) {
                return wb.Sheets[sheetName];
            }
        }
        return null;
    }

    _extractData(ws, requiredCols) {
        // Convert sheet to array of arrays
        const data = XLSX.utils.sheet_to_json(ws, { header: 1 });
        if (!data || data.length === 0) return [];

        let headerRowIdx = -1;
        let colMapping = {};

        // 1. Find Header Row (scan first 10 rows)
        for (let r = 0; r < Math.min(10, data.length); r++) {
            const rowValues = data[r];
            if (!rowValues) continue;

            let tempMapping = {};

            for (const [key, keywords] of Object.entries(requiredCols)) {
                for (let c = 0; c < rowValues.length; c++) {
                    const cellVal = this.normalizeString(rowValues[c]);
                    if (keywords.some(kw => cellVal.includes(kw))) {
                        tempMapping[key] = c;
                        break;
                    }
                }
            }

            // At least find 'item' and one amount
            if (Object.keys(tempMapping).length >= 2 && 'item' in tempMapping) {
                headerRowIdx = r;
                colMapping = tempMapping;
                break;
            }
        }

        if (headerRowIdx === -1) {
            console.warn("Could not identify headers");
            return [];
        }

        // 2. Extract Data
        const extractedData = [];
        for (let r = headerRowIdx + 1; r < data.length; r++) {
            const row = data[r];
            if (!row) continue;

            const rowData = {};
            let hasData = false;

            // Item Name
            if ('item' in colMapping) {
                const itemVal = row[colMapping['item']];
                if (itemVal) {
                    rowData['item'] = itemVal;
                    hasData = true;
                }
            }

            if (!hasData) continue;

            // Amounts
            for (const [key, colIdx] of Object.entries(colMapping)) {
                if (key === 'item') continue;
                let val = row[colIdx];

                // Clean number
                if (typeof val === 'number') {
                    rowData[key] = val;
                } else if (typeof val === 'string') {
                    // Remove commas
                    let cleanVal = val.replace(/,/g, "").trim();
                    // Handle (100) as -100
                    if (cleanVal.startsWith("(") && cleanVal.endsWith(")")) {
                        cleanVal = "-" + cleanVal.substring(1, cleanVal.length - 1);
                    }
                    const num = parseFloat(cleanVal);
                    rowData[key] = isNaN(num) ? 0 : num;
                } else {
                    rowData[key] = 0;
                }
            }
            extractedData.push(rowData);
        }
        return extractedData;
    }

    _processIncomeStatement(inputWb, outputWb) {
        const wsIn = this._findSheetByKeyword(inputWb, ["收支", "損益", "決算"]);
        const headers = ["科目名稱", "上年度決算數", "本年度預算數", "本年度決算數", "說明"];
        const outData = [headers];

        if (wsIn) {
            const data = this._extractData(wsIn, {
                "item": ["科目", "項目", "摘要", "名稱"],
                "this_year": ["本年", "本期", "決算"],
                "last_year": ["上年", "上期", "去年", "比較"],
                "budget": ["預算"]
            });

            data.forEach(row => {
                outData.push([
                    row.item || "",
                    row.last_year || 0,
                    row.budget || 0,
                    row.this_year || 0,
                    "系統轉檔"
                ]);
            });
        }

        const wsOut = XLSX.utils.aoa_to_sheet(outData);
        XLSX.utils.book_append_sheet(outputWb, wsOut, "收支決算表");
    }

    _processBalanceSheet(inputWb, outputWb) {
        const wsIn = this._findSheetByKeyword(inputWb, ["資產", "負債", "平衡"]);
        const headers = ["科目名稱", "上年度金額", "本年度金額", "說明"];
        const outData = [headers];

        if (wsIn) {
            const data = this._extractData(wsIn, {
                "item": ["科目", "項目", "摘要", "名稱"],
                "this_year": ["本年", "本期", "期末", "決算"],
                "last_year": ["上年", "上期", "期初", "去年"]
            });

            data.forEach(row => {
                outData.push([
                    row.item || "",
                    row.last_year || 0,
                    row.this_year || 0,
                    "系統轉檔"
                ]);
            });
        }

        const wsOut = XLSX.utils.aoa_to_sheet(outData);
        XLSX.utils.book_append_sheet(outputWb, wsOut, "資產負債表");
    }

    _processFundStatement(inputWb, outputWb) {
        // Keywords: 基金, 淨值, 變動
        const wsIn = this._findSheetByKeyword(inputWb, ["基金", "淨值", "變動"]);
        const headers = ["科目名稱", "期初餘額", "本期增加", "本期減少", "期末餘額"];
        const outData = [headers];

        if (wsIn) {
            const data = this._extractData(wsIn, {
                "item": ["科目", "項目", "名稱", "種類"],
                "start_balance": ["期初", "上期"],
                "increase": ["增加", "撥入"],
                "decrease": ["減少", "動支"],
                "end_balance": ["期末", "本期"]
            });

            data.forEach(row => {
                outData.push([
                    row.item || "",
                    row.start_balance || 0,
                    row.increase || 0,
                    row.decrease || 0,
                    row.end_balance || 0
                ]);
            });
        }

        const wsOut = XLSX.utils.aoa_to_sheet(outData);
        XLSX.utils.book_append_sheet(outputWb, wsOut, "基金收支表");
    }

    _processPropertyCatalog(inputWb, outputWb) {
        // Create empty placeholder if not found for now, or simple copy in future
        const headers = ["資產名稱", "取得日期", "取得成本", "本期折舊", "帳面價值"];
        const wsOut = XLSX.utils.aoa_to_sheet([headers]);
        XLSX.utils.book_append_sheet(outputWb, wsOut, "財產目錄");
    }
}
